Sources/AardvarkReveal/ARKRevealArchiveBuilder.m

```diff
@@ -0,0 +1,285 @@
+//
+//  Copyright 2020 Square, Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+#import "ARKRevealArchiveBuilder.h"
+
+#import <tar.h>
+
+
+struct header_posix_ustar {
+    char name[100];
+    char mode[8];
+    char uid[8];
+    char gid[8];
+    char size[12];
+    char mtime[12];
+    char checksum[8];
+    char typeflag[1];
+    char linkname[100];
+    char magic[6];
+    char version[2];
+    char uname[32];
+    char gname[32];
+    char devmajor[8];
+    char devminor[8];
+    char prefix[155];
+    char pad[12];
+};
+
+#define TAR_BLOCK_LENGTH 512
+
+
+NSString *const ARKRevealArchiveBuilderErrorDomain = @"ARKRevealArchiveBuilderErrorDomain";
+NSInteger const ARKRevealArchiveBuilderErrorCodeAlreadyComplete = 1;
+NSInteger const ARKRevealArchiveBuilderErrorCodeDataAllocationFailure = 2;
+NSInteger const ARKRevealArchiveBuilderErrorCodeHeaderWriteFailure = 3;
+
+
+@interface ARKRevealArchiveBuilder ()
+
+@property (nonatomic) NSMutableData *archiveData;
```

Comment:
I want to follow up here with having this write the archive data to disk, rather than storing it in memory. As is, this could potentially cause some OOM problems during the attachment generation. The complexity of that comes on the compression side, where it needs to page in data from disk while it's compressing it.